com.unity.render-pipelines.universal/Runtime/XR/XRSystemUniversal.cs

```diff
@@ -0,0 +1,69 @@
+using UnityEngine.Experimental.Rendering;
+
+namespace UnityEngine.Rendering.Universal
+{
+    internal static class XRSystemUniversal
+    {
+        // Prevent GC by keeping an array pre-allocated
+        static Matrix4x4[] s_projMatrix = new Matrix4x4[2];
+
+        internal static void BeginLateLatching(Camera camera, XRPassUniversal xrPass)
+        {
+#if ENABLE_VR && ENABLE_XR_MODULE
+            XR.XRDisplaySubsystem xrDisplay = XRSystem.GetActiveDisplay();
+
+            if (xrDisplay != null && xrPass.viewCount == 2) // multiview only
+            {
+                xrDisplay.BeginRecordingIfLateLatched(camera);
+                xrPass.isLateLatchEnabled = true;
+            }
+#endif
+        }
+
+        internal static void EndLateLatching(Camera camera, XRPassUniversal xrPass)
+        {
+#if ENABLE_VR && ENABLE_XR_MODULE
+            XR.XRDisplaySubsystem xrDisplay = XRSystem.GetActiveDisplay();
+
+            if (xrDisplay != null && xrPass.isLateLatchEnabled)
+            {
+                xrDisplay.EndRecordingIfLateLatched(camera);
+                xrPass.isLateLatchEnabled = false;
+            }
+#endif
+        }
+
+        internal static void UnmarkShaderProperties(CommandBuffer cmd, XRPassUniversal xrPass)
+        {
+#if ENABLE_VR && ENABLE_XR_MODULE
+            if (xrPass.isLateLatchEnabled && xrPass.hasMarkedLateLatch)
+            {
+                cmd.UnmarkLateLatchMatrix(CameraLateLatchMatrixType.View);
+                cmd.UnmarkLateLatchMatrix(CameraLateLatchMatrixType.InverseView);
+                cmd.UnmarkLateLatchMatrix(CameraLateLatchMatrixType.ViewProjection);
+                cmd.UnmarkLateLatchMatrix(CameraLateLatchMatrixType.InverseViewProjection);
+                xrPass.hasMarkedLateLatch = false;
+            }
+#endif
+        }
+
+        internal static void MarkShaderProperties(CommandBuffer cmd, XRPassUniversal xrPass, bool renderIntoTexture)
+        {
+#if ENABLE_VR && ENABLE_XR_MODULE
+            if (xrPass.isLateLatchEnabled && xrPass.canMarkLateLatch)
```

Comment:
Previously, MarkLateLatchShaderProperties is called in UpdateGPUViewAndProjectionMatrices() under the condition that cameraData.xr.singlePassEnabled is true. Currently, that's no longer the case. We need to restore it back as before:

    internal void UpdateGPUViewAndProjectionMatrices()
        {
     ...
            if (cameraData.xr.singlePassEnabled)
            {
               ...
                if (cameraData.xr.canMarkLateLatch)
                    MarkLateLatchShaderProperties(cmd, ref cameraData);
            }
        }

Also, there's the plan to change xrPass.isLateLatchEnabled based on return value of BeginRecordingIfLateLatched(). In that case, xrPass.isLateLatchEnabled would strictly be based on XRDisplay's capabilities rather than being multiview or not (currently). Thus, it's better to also add the additional cameraData.xr.singlePassEnabled check here.

Fyi, thanks for adding this XRSystemUniversal container!

Reply:
Actually, it is redundant.
isLateLatchEnabled can only be set to true if singlePassEnabled is true (see BeginLateLatching).
So whenever isLateLatchEnabled is true, we can be sure that singlePassEnabled is also true.

Reply:
You didn't finish reading my last comment: 
"Also, there's the plan to change xrPass.isLateLatchEnabled based on return value of BeginRecordingIfLateLatched(). In that case, xrPass.isLateLatchEnabled would strictly be based on XRDisplay's capabilities rather than being multiview or not (currently). Thus, it's better to also add the additional cameraData.xr.singlePassEnabled check here."

Thus, we should keep my previous logic the same during this refactor so that engine's BeginRecordingIfLateLatch changes would just work without further changes to URP.

Reply:
Upon more discussion, we can leave this line as is, and later we can add vendor-specific policy decisions to xrUniversal.canMarkLateLatch (i.e. whether a vendor (Oculus, MS) allows late latching in singlePass or multipass or both). With new BeginRecordingIfLateLatched changes, isLateLatchEnabled will still be based on the vendor capability flag.